projects/chaos/src/sequencer.c

```diff
@@ -0,0 +1,365 @@
+  FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_RESET, sequencer_state_drive);
+}
+
+// FSM Transitions
+static void prv_sequencer_state_emergency(FSM *fsm, const Event *e, void *context) {
+  (void)fsm;
+  (void)e;
+  prv_sequencer_storage_update((SequencerStorage *)context,
+                               (SequencerEventPair *)s_emergency_events,
+                               SIZEOF_ARRAY(s_emergency_events));
+}
+
+static void prv_sequencer_state_idle(FSM *fsm, const Event *e, void *context) {
+  (void)fsm;
+  (void)e;
+  prv_sequencer_storage_update((SequencerStorage *)context, (SequencerEventPair *)s_idle_events,
+                               SIZEOF_ARRAY(s_idle_events));
+}
+
+static void prv_sequencer_state_charge(FSM *fsm, const Event *e, void *context) {
+  (void)fsm;
+  (void)e;
+  prv_sequencer_storage_update((SequencerStorage *)context, (SequencerEventPair *)s_charge_events,
+                               SIZEOF_ARRAY(s_charge_events));
+}
+
+static void prv_sequencer_state_drive(FSM *fsm, const Event *e, void *context) {
+  (void)fsm;
+  (void)e;
+  prv_sequencer_storage_update((SequencerStorage *)context, (SequencerEventPair *)s_drive_events,
+                               SIZEOF_ARRAY(s_drive_events));
+}
+
+// Public interface
+void sequencer_init(void) {
+  memset(&s_storage, 0, sizeof(s_storage));
+  fsm_state_init(sequencer_state_emergency, prv_sequencer_state_emergency);
+  fsm_state_init(sequencer_state_idle, prv_sequencer_state_idle);
+  fsm_state_init(sequencer_state_charge, prv_sequencer_state_charge);
+  fsm_state_init(sequencer_state_drive, prv_sequencer_state_drive);
+  fsm_init(&s_sequencer_fsm, "SequenceFSM", &sequencer_state_idle, &s_storage);
+}
+
+StatusCode sequencer_publish_next_event(const Event *previous_event) {
+  // Filter to only handled events. Those in the range
+  // |(NUM_CHAOS_EVENTS_CAN, NUM_CHAOS_EVENTS_FSM)|
+  // in addition to the emergency sequence messages.
+  if (previous_event->id <= NUM_CHAOS_EVENTS_CAN &&
+      !(previous_event->id == CHAOS_EVENT_SEQUENCE_EMERGENCY ||
+        previous_event->id == CHAOS_EVENT_SEQUENCE_EMERGENCY_DONE)) {
+    // Not in the range of handled events. This isn't necessarily invalid but won't be handled so
+    // return unknown.
+    return status_code(STATUS_CODE_UNKNOWN);
+  } else if (previous_event->id == CHAOS_EVENT_NO_OP) {
+    // Skip no-op messages.
+    return STATUS_CODE_OK;
+  }
+
+  // Try to transition if the event is a sequence transition.
+  if (previous_event->id == CHAOS_EVENT_SEQUENCE_IDLE ||
+      previous_event->id == CHAOS_EVENT_SEQUENCE_CHARGE ||
+      previous_event->id == CHAOS_EVENT_SEQUENCE_DRIVE ||
+      previous_event->id == CHAOS_EVENT_SEQUENCE_EMERGENCY ||
+      previous_event->id == CHAOS_EVENT_SEQUENCE_RESET) {
+    if (!fsm_process_event(&s_sequencer_fsm, previous_event)) {
+      LOG_DEBUG("Err failed transition: %u %u\n", previous_event->id, previous_event->data);
```

Comment:
We technically have other `LOG_` types if you want to use those

Reply:
Done.